### DISCUSSION TITLE
Factory/family capsule rebuild

### DISCUSSION THREAD

**Discussion by busslina** [GregoryConrad/rearch-dart#48] [Q&A] [answered]
Let's say I have a `Set<String>` for every websocket connection (In my real world app it correspond to passwords id's). This is a base code:
```dart
_Controller Function(ConnectionKey) passwordIdsControllerFamily(CapsuleHandle use) {
  final controllers = use.value(<ConnectionKey, _Controller>{});

  return (connectionkey) {
    final controller = controllers[connectionkey];

    if (controller != null) {
      return controller;
    }

    final newController = controllers[connectionkey] = _Controller();

    newController.load();

    return newController;
  };
}

class _Controller {
  AsyncValue<Set<String>> ids = const AsyncLoading(None());

  Future<void> load() => throw ('Not implemented');
}
```

So the issue that I'm thinking of would be when I receive an event for an added/removed id for a unique websocket connection.
I can think of two options to notify a change to my Table widget:
1) Brute force: To rebuild the whole `passwordIdsControllerFamily` capsule. This way I will rebuild also widgets that only use other websocket connection ids.
2) Custom notifiers. Using notifiers for every `_Controller`. That way is more selective but I add extra logic outside the ReArch mechanisms.

What do you think about this?

**Comment by busslina** [comment]
Okay, I remebered the reduce rebuild doc section. I'm gonna take a look again over that. Seems to the solution.

Anyway, would be ok for you to rebuild the whole capsule in this scenario?

**Comment by GregoryConrad** [comment, marked as answer]
short answer: it depends.

How many downstream capsules are using the factory? If the answer is none (I.e., you’re doing container.read of the factory from a webserver), then you can probably do manual listeners in the controller. If you do have some downstream capsules, you may want to rely on ReArch’s propagation (and can .map() the capsules into something else to avoid some widget rebuilds).

idk if it’s possible here, but your controllers should likely be entirely immutable. Capsule data tends to bode better when immutable.

An fyi: as a best practice your code can be simplified with a hashmap.putIfAbsent instead of a null check in the factory.

also, I’ll review your other discussion post when I get a chance I’ve been really busy the past few days

**Reply by busslina** [reply]
I wanted to call it `factory` but I did `family` hehe (dirty local cache).

> How many downstream capsules are using the factory?

Only one capsule is using it, but that capsule is not "reactive". It just provide the readed `controllerA` to another just created `controllerB` which will call eventually on `controllerA` remote CRUD events received via websockets.
